doc/internals/file-system.md

```diff
@@ -0,0 +1,206 @@
+Unix style name strings: "REMOTE:/home/tom/abc.lisp" while the "LOCAL"
+host uses LispM style name strings: "LOCAL:>home>tom>abc.lisp".
+
+For a file system that resides on a local disk or disk partition, the
+host object must have a block device object so that read and write
+requests can be mapped to the appropriate disk location. Removable
+media, such as USB mass storage devices, also provide block device
+objects.
+
+When Mezzano boots, it enumerates all of the available disks and disk
+partitions. A list of these objects can be obtained by calling:
+
+    mezzano.supervisor:all-disks () => List of disks and partitions
+
+These disks and partitions also support the block-device APIs:
+
+    mezzano.disk:block-device-sector-size (disk) => <sector size in bytes>
+
+    mezzano.disk:block-device-read (disk lba n-sectors buffer &key (offset 0)) => T on success, NIL and error otherwise
+
+    mezzano.disk:block-device-write (disk lba n-sectors buffer &key (offset 0)) => T on success, NIL and error otherwise
+
+    mezzano.disk:block-device-flush (disk) => T on success, NIL and error otherwise
+
+A list of block devices can obtained by calling:
+
+    mezzano.disk:all-block-devices () => List of block devices
+
+All of the local disks and partitions enumerated on boot and returned
+by mezzano.supervisor:all-disks are also returned by
+mezzano.disk:all-block-devices. Additional block devices can be
+registered by:
+
+    mezzano.disk:register-block-device (device) => Updated list of block devices
+
+And unregistered by:
+
+    mezzano.disk:unregister-block-device (device) => Updated list of block devices
+
+To create a new file system on a running system, create a host object
+of the appropriate type with the desired host name and the desired
+block device object. For example, the type could be fat32, the host
+name could be "HOME", and the block device  selected from the results of
+(mezzano.disk:all-block-devices). For this example, the host name/host
+object mapping would be setup by
+
+    (setf (mezzano.file-system:find-host  "HOME") <fat32 host object>)
+
+If the system is rebooted without saving the current state using:
+
+    mezzano.supervisor:snapshot () => <ignored>
+
+then the host object will be lost and will need to be recreated and
+mapped to the appropriate host name and associated with the
+appropriate block device object.
+
+However, if the system is saved, when the system is rebooted, the host
+object and host name mapping will still exist and the host will be
+re-associated with the block device via the following mechanism.
+
+There is a class "file-host-mount-mixin" which includes a "mount-args"
+slot which is accessed via an accessor function file-host-mount-args
+and defines a generic function:
+
+    mount (host) => T on success, NIL otherwise
+
+and includes a default mount method which does nothing and returns T
+(success).
+
+All file system host classes include file-host-mount-mixin and when a
```

Comment:
I'm not sure about requiring that non-block-based file systems subscribe to the mount protocol.
For network-based file systems (remote & a hypothetical NFS implementation) it seems fine, but for the local file system and for logical hosts (which don't really exist as real file systems at all) it doesn't make much sense. They exist from the moment their host objects are created and are destroyed when they get GC'd.
I don't know what the behaviour here should be either, so lets go with this for now. It can always be changed later.